segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -0,0 +1,308 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.pravega.segmentstore.server.containers;
+
+import com.google.common.base.Preconditions;
+import com.google.common.util.concurrent.AbstractIdleService;
+import io.pravega.common.Timer;
+import io.pravega.common.concurrent.Futures;
+import io.pravega.common.util.BufferView;
+import io.pravega.common.util.ByteArraySegment;
+import io.pravega.segmentstore.contracts.AttributeUpdate;
+import io.pravega.segmentstore.contracts.AttributeUpdateType;
+import io.pravega.segmentstore.contracts.ReadResult;
+import io.pravega.segmentstore.contracts.ReadResultEntry;
+import io.pravega.segmentstore.server.ContainerEventProcessor;
+import io.pravega.segmentstore.server.DirectSegmentAccess;
+import io.pravega.segmentstore.server.SegmentStoreMetrics;
+import lombok.Cleanup;
+import lombok.Data;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionException;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+/**
+ * Implementation for {@link ContainerEventProcessor}. This class stores a map of {@link ContainerEventProcessor.EventProcessor}
+ * identified by name. Once this service is started, it will continuously perform processing iterations tailing the
+ * internal Segments of the registered {@link ContainerEventProcessor.EventProcessor}s (every ITERATION_SLEEP ms) and
+ * (safely) invoking their respective handler functions. This class durably stores the point at which the last successful
+ * events has been processed in the form of an attribute in the internal Segment (i.e., LAST_PROCESSED_EVENT_OFFSET).
+ * That is, before reading new events this attribute is read and after a successful processing of events this attribute
+ * is updated with the value of the next position to read. This class also truncates the internal Segments of the registered
+ * {@link ContainerEventProcessor.EventProcessor} based on that attribute.
+ */
+@Slf4j
+public final class ContainerEventProcessorImpl extends AbstractIdleService implements ContainerEventProcessor {
+
+    private static final UUID LAST_PROCESSED_EVENT_OFFSET = new UUID(0, 0);
```

Comment:
Let's make this a `Core` Attribute and put it in `Attributes`. Core attributes are always stored in memory and are never flushed to Storage. I expect this to be a frequently accessed one so let's do it.

Reply:
I have removed this attribute altogether. As we discussed, suing the `getOffset()` and `getLength()` information of a Segment we can achieve the same without adding a new attribute.